packages/router/utils.ts

```diff
@@ -783,6 +841,68 @@ export const json: JsonFunction = (data, init = {}) => {
   });
 };
 
+export class DeferredData {
+  private pendingKeys: Set<string> = new Set<string>();
+  private cancelled: boolean = false;
+  private subscriber?: (key: string, data: any) => void = undefined;
+  data: RouteData = {};
+
+  constructor(data: Record<string, any>) {
+    Object.entries(data).forEach(([key, value]) => {
+      // Store all data in our internal copy and track promise keys
+      this.data[key] = value;
+      if (value instanceof Promise) {
+        this.pendingKeys.add(key);
+        value.then(
+          (data) => this.onSettle(key, null, data),
+          (error) => this.onSettle(key, error)
+        );
+      }
+    });
+  }
+
+  private onSettle(key: string, error: any, data?: any) {
+    if (this.cancelled) {
+      return;
+    }
+    this.pendingKeys.delete(key);
+    let value = error ? new DeferredError(error) : data;
+    this.data[key] = value;
+    this.subscriber?.(key, value);
+  }
+
+  subscribe(fn: (key: string, data: any) => void) {
+    this.subscriber = fn;
+  }
+
+  cancel() {
+    this.cancelled = true;
+    this.pendingKeys.forEach((v, k) => this.pendingKeys.delete(k));
+  }
+
+  get done() {
+    return this.pendingKeys.size === 0;
+  }
+}
+
+/**
+ * @private
+ * Utility class we use to hold deferred promise rejection values
+ */
+export class DeferredError extends Error {}
+
+/**
+ * Check if the given error is a DeferredError generated from a deferred()
+ * promise rejection
+ */
+export function isDeferredError(e: any): e is DeferredError {
+  return e instanceof DeferredError;
```

Comment:
This is so we know if the data on `loaderData` was from a promise rejection and we should render it in the `<Deferred>` `errorBoundary` component